resteasy-jaxrs/src/main/java/org/jboss/resteasy/core/ServerResponseWriter.java

```diff
@@ -403,49 +418,61 @@ public int compareTo(SortableMediaType o)
          {
             return 1;
          }
+         if (this.writerType == o.writerType) return 0;
+         if (this.writerType == null) return -1;
+         if (o.writerType == null) return 1;
+         if (this.writerType.isAssignableFrom(o.writerType)) return -1;
+         if (o.writerType.isAssignableFrom(this.writerType)) return 1;
          return 0;
       }
    }
    
+   
+   private static SortableMediaType mostSpecific(MediaType m1, MediaType m2)
+   {
+      Class<?> wt1 = m1 instanceof SortableMediaType ? ((SortableMediaType)m1).writerType : null;
+      Class<?> wt2 = m2 instanceof SortableMediaType ? ((SortableMediaType)m2).writerType : null;
+      return mostSpecific(m1, wt1, m2, wt2);
+   }
+   
    /**
     * m1, m2 are compatible
     */
-   private static SortableMediaType mostSpecific(MediaType m1, MediaType m2)
+   private static SortableMediaType mostSpecific(MediaType m1, Class<?> wt1, MediaType m2, Class<?> wt2)

```

Comment:
Should this method not be the S fucntion defined by spec section 3.7.2 ?

- if m1="*/xml" and m2="application/*" => the method wil return "application/*"  instead  of "application/xml;d=2"
- if m1="text/*;q=0.5;" and m2="text/html;qs=0.8" => the method will return "text/html;qs=0.8" instead of "text/html;q=0.5;qs=0.8;d=1"